subprojects/core/src/integTest/groovy/org/gradle/api/provider/PropertyStateIntegrationTest.groovy

```diff
@@ -0,0 +1,372 @@
+                    outputFiles.set(getProject().files());
+                }
+
+                @Input
+                public boolean getEnabled() {
+                    return enabled.get();
+                }
+                
+                public void setEnabled(PropertyState<Boolean> enabled) {
+                    this.enabled = enabled;
+                }
+
+                public void setEnabled(boolean enabled) {
+                    this.enabled.set(enabled);
+                }
+
+                @OutputFiles
+                public ConfigurableFileCollection getOutputFiles() {
+                    return outputFiles.get();
+                }
+
+                public void setOutputFiles(PropertyState<ConfigurableFileCollection> outputFiles) {
+                    this.outputFiles = outputFiles;
+                }
+
+                public void setOutputFiles(ConfigurableFileCollection outputFiles) {
+                    this.outputFiles.set(outputFiles);
+                }
+                
+                @TaskAction
+                public void resolveValue() throws IOException {
+                    if (getEnabled()) {
+                        for (File outputFile : getOutputFiles()) {
+                            writeFile(outputFile, "$OUTPUT_FILE_CONTENT");
+                        }
+                    }
+                }
+                
+                private void writeFile(File destination, String content) throws IOException {
+                    BufferedWriter output = null;
+                    try {
+                        output = new BufferedWriter(new FileWriter(destination));
+                        output.write(content);
+                    } finally {
+                        if (output != null) {
+                            output.close();
+                        }
+                    }
+                }
+            }
+        """
+    }
+
+    private String pluginWithExtensionMapping(Closure taskCreation) {
+        """
+            apply plugin: MyPlugin
+            
+            pluginConfig {
+                enabled = true
+                outputFiles = files('${normaliseFileSeparators(customOutputFile.canonicalPath)}')
+            }
+
+            class MyPlugin implements Plugin<Project> {
+                void apply(Project project) {
+                    def extension = project.extensions.create('pluginConfig', MyExtension, project)
+
+                    ${taskCreation()}
+                }
+            }
+
+            class MyExtension {
+                private final PropertyState<Boolean> enabled
+                private final PropertyState<FileCollection> outputFiles
+
+                MyExtension(Project project) {
+                    enabled = project.property(Boolean)
+                    outputFiles = project.property(FileCollection)
+                }
+
+                PropertyState<Boolean> getEnabled() {
```

Comment:
We need to use the same pattern for both tasks and extensions. This would mean the getter should probably return the value and there would be another method to query a property as a provider.